`seq!` for tuples errors when capturing locally-defined parsers.

### Please complete the following tasks

- [X] I have searched the [discussions](https://github.com/winnow-rs/winnow/discussions)
- [X] I have searched the [open](https://github.com/winnow-rs/winnow/issues) and [rejected](https://github.com/winnow-rs/winnow/issues?q=is%3Aissue+label%3AS-wont-fix+is%3Aclosed) issues

### rust version

rustc 1.77.0 (aedd173a2 2024-03-17)

### winnow version

0.6.5

### Minimal reproducible code

```rust
use winnow::combinator::seq;
use winnow::prelude::*;

pub fn doesnt_work(input: &mut &str) -> PResult<(i32, i32)> {
    let mut foo = "foo".map(|_| 10);
    let mut bar = "bar".map(|_| 20);
    seq!(foo, bar).parse_next(input)
}
```


### Steps to reproduce the bug with the above code

`cargo check`

### Actual Behaviour

```
error[E0525]: expected a closure that implements the `FnMut` trait, but this closure only implements `FnOnce`
  --> good-crates/parser/examples/mcve.rs:18:5
   |
18 |     seq!(foo, bar).parse_next(input)
   |     ^^^^^---^^^^^^
   |     |    |
   |     |    closure is `FnOnce` because it moves the variable `foo` out of its environment
   |     this closure implements `FnOnce`, not `FnMut`
   |     the requirement to implement `FnMut` derives from here
   |     required by a bound introduced by this call
   |
   = note: required for `{closure@/home/alona/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winnow-0.6.5/src/macros/seq.rs:88:44: 88:64}` to implement `winnow::Parser<&str, (i32, i32), ContextError>`
note: required by a bound in `trace`
  --> /home/alona/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winnow-0.6.5/src/combinator/debug/mod.rs:45:18
   |
43 | pub fn trace<I: Stream, O, E>(
   |        ----- required by a bound in this function
44 |     name: impl crate::lib::std::fmt::Display,
45 |     parser: impl Parser<I, O, E>,
   |                  ^^^^^^^^^^^^^^^ required by this bound in `trace`
   = note: this error originates in the macro `$crate::seq` which comes from the expansion of the macro `seq` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0525`.
```

### Expected Behaviour

This should compile.

If `seq!` is used to make a struct (instead of a tuple) it does compile.

```rust
pub struct Res {
    pub a: i32,
    pub b: i32,
}

pub fn does_work(input: &mut &str) -> PResult<Res> {
    let mut foo = "foo".map(|_| 10);
    let mut bar = "bar".map(|_| 20);
    seq! { Res{ a: foo, b: bar } }.parse_next(input)
}
```

works fine.

### Additional Context

Manually expanding the `seq!` macro (with RA) for the broken example gives:

```rust
pub fn doesnt_work(input: &mut &str) -> PResult<(i32, i32)> {
    let mut foo = "foo".map(|_| 10);
    let mut bar = "bar".map(|_| 20);
    {
        winnow::combinator::trace("tuple", move |input: &mut _| {
            use winnow::Parser;
            (foo, bar).map(|t| ((t.0), (t.1))).parse_next(input)
        })
    }
    .parse_next(input)
}
```

This gives a more descriptive error:

```

error[E0525]: expected a closure that implements the `FnMut` trait, but this closure only implements `FnOnce`
  --> good-crates/parser/examples/mcve.rs:19:44
   |
19 |           winnow::combinator::trace("tuple", move |input: &mut _| {
   |           -------------------------          -^^^^^^^^^^^^^^^^^^^
   |           |                                  |
   |  _________|__________________________________this closure implements `FnOnce`, not `FnMut`
   | |         |
   | |         required by a bound introduced by this call
20 | |             use winnow::Parser;
21 | |             (foo, bar).map(|t| ((t.0), (t.1))).parse_next(input)
   | |              --- closure is `FnOnce` because it moves the variable `foo` out of its environment
22 | |         })
   | |_________- the requirement to implement `FnMut` derives from here
   |
   = note: required for `{closure@good-crates/parser/examples/mcve.rs:19:44: 19:64}` to implement `winnow::Parser<&str, (i32, i32), ContextError>`
note: required by a bound in `trace`
  --> /home/alona/.cargo/registry/src/index.crates.io-6f17d22bba15001f/winnow-0.6.5/src/combinator/debug/mod.rs:45:18
   |
43 | pub fn trace<I: Stream, O, E>(
   |        ----- required by a bound in this function
44 |     name: impl crate::lib::std::fmt::Display,
45 |     parser: impl Parser<I, O, E>,
   |                  ^^^^^^^^^^^^^^^ required by this bound in `trace`

For more information about this error, try `rustc --explain E0525`.
```

Manually expanding the example that does work gives:

```rust
pub fn does_work(input: &mut &str) -> PResult<Res> {
    let mut foo = "foo".map(|_| 10);
    let mut bar = "bar".map(|_| 20);
    {
        winnow::combinator::trace("Res", move |input: &mut _| {
            use winnow::Parser;
            let a = foo.parse_next(input)?;
            let b = bar.parse_next(input)?;
            #[allow(clippy::redundant_field_names)]
            Ok(Res { a: a, b: b })
        })
    }
    .parse_next(input)
}
```

I think the reason this works is that `foo` and `bar` don't need to be captured by value here, as `.parse_next` takes `&mut self`.

Adding `.by_ref()` to the tuple example version may fix it.

--- a/src/macros/seq.rs
+++ b/src/macros/seq.rs
@@ -44,7 +44,7 @@
 ///
 /// // Or parse into tuples
 /// fn point(input: &mut &[u8]) -> PResult<(u32, u32)> {
-///     let num = dec_uint::<_, u32, ContextError>;
+///     let mut num = dec_uint::<_, u32, ContextError>;
 ///     seq!(num, _: (space0, b',', space0), num).parse_next(input)
 /// }
 ///
@@ -89,27 +89,28 @@ macro_rules! seq {
         $crate::combinator::trace(stringify!($($name)::*), move |input: &mut _| {
             $crate::seq_parse_tuple_fields!(
                 ( $($fields)* );
-            ).map(|t| {
-                $crate::seq_init_tuple_fields!(
-                    ( $($fields)* );
-                    ( t.0, t.1, t.2, t.3, t.4, t.5, t.6, t.7, t.8, t.9, t.10, t.11, t.12, t.13, t.14, t.15, t.16, t.17, t.18, t.19, t.20 );
-                    $($name)::* ;
-                )
-            }).parse_next(input)
+                ( _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20 );
+                input;
+            );
+            Ok($crate::seq_init_tuple_fields!(
+                ( $($fields)* );
+                ( _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20 );
+                $($name)::*;
+            ))
         })
     };
     (( $($fields: tt)* )) => {
         $crate::combinator::trace("tuple", move |input: &mut _| {
-            use $crate::Parser;
             $crate::seq_parse_tuple_fields!(
                 ( $($fields)* );
-            ).map(|t| {
-                $crate::seq_init_tuple_fields!(
-                    ( $($fields)* );
-                    ( t.0, t.1, t.2, t.3, t.4, t.5, t.6, t.7, t.8, t.9, t.10, t.11, t.12, t.13, t.14, t.15, t.16, t.17, t.18, t.19, t.20 );
-                    ;
-                )
-            }).parse_next(input)
+                ( _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20 );
+                input;
+            );
+            Ok($crate::seq_init_tuple_fields!(
+                ( $($fields)* );
+                ( _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20 );
+                ;
+            ))
         })
     };
     ($($fields: tt)*) => {
@@ -176,47 +177,29 @@ macro_rules! seq_parse_struct_fields {
 #[doc(hidden)]
 macro_rules! seq_parse_tuple_fields {
     (
-        ( _ : $head_parser: expr, $($fields: tt)* );
-        $($sequenced: tt)*
+        ( $(_ :)? $head_parser: expr, $($fields: tt)* );
+        ( $unnamed1: ident, $($unnamed: ident),* );
+        $input: ident;
     ) => {
+        let $unnamed1 = $crate::Parser::parse_next(&mut $head_parser, $input)?;
         $crate::seq_parse_tuple_fields!(
             ( $($fields)* );
-            $($sequenced)* $head_parser.void(),
+            ( $($unnamed),* );
+            $input ;
         )
     };
     (
-        ( _ : $head_parser: expr );
-        $($sequenced: tt)*
+        ( $(_ :)? $head_parser: expr );
+        ( $unnamed1: ident, $($unnamed: ident),* );
+        $input: ident;
     ) => {
-        $crate::seq_parse_tuple_fields!(
-            ();
-            $($sequenced)* $head_parser.void(),
-        )
+        let $unnamed1 = $crate::Parser::parse_next(&mut $head_parser, $input)?;
     };
     (
-        ( $head_parser: expr, $($fields: tt)* );
-        $($sequenced: tt)*
-    ) => {
-        $crate::seq_parse_tuple_fields!(
-            ( $($fields)* );
-            $($sequenced)* $head_parser,
-        )
-    };
-    (
-        ( $head_parser: expr );
-        $($sequenced: tt)*
-    )=> {
-        $crate::seq_parse_tuple_fields!(
-            ();
-            $($sequenced)* $head_parser,
-        )
-    };
-    (
-        ();
-        $($sequenced: tt)*
-    ) => {
-        ($($sequenced)*)
-    };
+        ( $(,)? );
+        ( $($unnamed: ident),* );
+        $input: expr;
+    ) => {};
 }
 
 #[macro_export]
@@ -288,62 +271,63 @@ macro_rules! seq_init_struct_fields {
 macro_rules! seq_init_tuple_fields {
     (
         ( _ : $head_parser: expr, $($fields: tt)* );
-        ( $head_arg: expr, $($args: expr),* );
-        $($name: ident)::* ;
+        ( $unnamed1: ident, $($unnamed: ident),* );
+        $($name: ident)::*;
         $($inits: tt)*
     ) => {
         $crate::seq_init_tuple_fields!(
             ( $($fields)* );
-            ( $($args),* );
+            ( $($unnamed),* );
             $($name)::* ;
             $($inits)*
         )
     };
     (
         ( _ : $head_parser: expr );
-        ( $head_arg: expr, $($args: expr),* );
-        $($name: ident)::* ;
+        ( $unnamed1: ident, $($unnamed: ident),* );
+        $($name: ident)::*;
         $($inits: tt)*
     ) => {
         $crate::seq_init_tuple_fields!(
             ();
-            ( $($args),* );
+            ( $($unnamed),* );
             $($name)::* ;
             $($inits)*
         )
     };
     (
         ( $head_parser: expr, $($fields: tt)* );
-        ( $head_arg: expr, $($args: expr),* );
-        $($name: ident)::* ;
+        ( $unnamed1: ident, $($unnamed: ident),* );
+        $($name: ident)::*;
         $($inits: tt)*
-    ) => {
+    ) =>
+    {
         $crate::seq_init_tuple_fields!(
             ( $($fields)* );
-            ( $($args),* );
+            ( $($unnamed),* );
             $($name)::* ;
-            $($inits)* $head_arg,
+            $($inits)* $unnamed1,
         )
     };
     (
         ( $head_parser: expr );
-        ( $head_arg: expr, $($args: expr),* );
-        $($name: ident)::* ;
+        ( $unnamed1: ident, $($unnamed: ident),* );
+        $($name: ident)::*;
         $($inits: tt)*
     ) => {
         $crate::seq_init_tuple_fields!(
             ();
-            ( $($args),* );
+            ( $($unnamed),* );
             $($name)::* ;
-            $($inits)* $head_arg
+            $($inits)* $unnamed1,
         )
     };
     (
-        ();
-        ( $($args: expr),* );
-        $($name: ident)::* ;
-        $($inits: expr),* $(,)?
+        ( $(,)? );
+        ( $unnamed1: ident, $($unnamed: ident),* );
+        $($name: ident)::*;
+        $($inits: tt)*
     ) => {
-        $($name)::*( $($inits,)* )
+        $($name)::* ( $($inits)* )
     };
 }

--- a/src/macros/test.rs
+++ b/src/macros/test.rs
@@ -438,3 +438,24 @@ fn seq_tuple_no_parens() {
         .parse_next(input)
     }
 }
+
+#[test]
+fn seq_tuple_borrow() {
+    #![allow(dead_code)]
+
+    #[derive(Debug, PartialEq)]
+    struct Point(u32, u32);
+
+    fn parser(input: &mut &str) -> PResult<Point> {
+        let mut dec_uint = digit0.parse_to();
+        seq! {
+            Point(
+                dec_uint,
+                _: ',',
+                dec_uint,
+                _: empty
+            )
+        }
+        .parse_next(input)
+    }
+}

--- a/src/macros/test.rs
+++ b/src/macros/test.rs
@@ -1,4 +1,5 @@
 use crate::ascii::dec_uint;
+use crate::ascii::digit0;
 use crate::combinator::dispatch;
 use crate::combinator::empty;
 use crate::combinator::fail;
@@ -222,6 +223,30 @@ fn seq_enum_struct_variant() {
         mul.parse_peek(&b"3 * 4"[..]),
         Ok((&b""[..], Expr::Mul(3, 4))),
     );
+}
+
+#[test]
+fn seq_struct_borrow() {
+    #![allow(dead_code)]
+
+    #[derive(Debug, PartialEq)]
+    struct Point {
+        x: u32,
+        y: u32,
+    }
+
+    fn parser(input: &mut &str) -> PResult<Point> {
+        let mut dec_uint = digit0.parse_to();
+        seq! {
+            Point {
+                x: dec_uint,
+                _: ',',
+                y: dec_uint,
+                _: empty
+            }
+        }
+        .parse_next(input)
+    }
 }
 
 #[test]

